### ISSUE TITLE
Null objects do not extend their original class

### ISSUE THREAD

**Issue by trickleup** [Ocramius/ProxyManager#300]
It seems to me that classes with PHP7 return types fail.

```
PHP Fatal error:  Declaration of ProxyManagerGeneratedProxy\__PM__\MyClass::queryUsers() must be compatible with MyClass::queryUsers(): array in /Users/johan/Projects/sparkcentral/auth-context/vendor/ocramius/proxy-manager/src/ProxyManager/GeneratorStrategy/EvaluatingGeneratorStrategy.php(63) : eval()'d code on line 3
```

Is there a workaround/fix for this issue?

**Comment by trickleup** [comment]
Also, the same seems to be the case with scalar type hints in method declarations.

**Comment by Ocramius** [comment]
@trickleup ProxyManager v2 fully supports return type hints: can you check what version you have installed?

**Comment by michaelmoussa** [comment]
@Ocramius I'm having (_what I think is_) the same issue, only with the `Null Object Proxy`. See example case below:

```
$ cat test.php
<?php

declare (strict_types = 1);

require __DIR__ . '/vendor/autoload.php';

echo \ProxyManager\Version::getVersion() . "\n";

class Foo
{
}

function bar() : Foo
{
    return (new \ProxyManager\Factory\NullObjectFactory())->createProxy(Foo::class);
}

bar();
```

```
$ php test.php
2.0.1@6c89b7bd6039d8047b1473e2074cb56baa4bc15d
PHP Fatal error:  Uncaught TypeError: Return value of bar() must be an instance of Foo, instance of ProxyManagerGeneratedProxy\__PM__\Foo\Generatede07f73e523c2fb4772e096fa7615d794 returned in /Users/mmoussa/Code/pm-test/test.php:15
```

The generated code looks like this:

```php
namespace ProxyManagerGeneratedProxy\__PM__\Foo;

class Generatede07f73e523c2fb4772e096fa7615d794 implements \ProxyManager\Proxy\NullObjectInterface
{

    private static $signaturee07f73e523c2fb4772e096fa7615d794 = 'YTozOntzOjk6ImNsYXNzTmFtZSI7czozOiJGb28iO3M6NzoiZmFjdG9yeSI7czozODoiUHJveHlNYW5hZ2VyXEZhY3RvcnlcTnVsbE9iamVjdEZhY3RvcnkiO3M6MTk6InByb3h5TWFuYWdlclZlcnNpb24iO3M6NDY6IjIuMC4xQDZjODliN2JkNjAzOWQ4MDQ3YjE0NzNlMjA3NGNiNTZiYWE0YmMxNWQiO30=';

    /**
     * Constructor for null object initialization
     */
    public static function staticProxyConstructor()
    {
        static $reflection;

        $reflection = $reflection ?: $reflection = new \ReflectionClass(__CLASS__);
        $instance = (new \ReflectionClass(get_class()))->newInstanceWithoutConstructor();

        return $instance;
    }
}
```

Since `Generatede07f73e523c2fb4772e096fa7615d794` doesn't extend `Foo`, it will obviously fail the return type check.

If I add this: `$classGenerator->setExtendedClass($originalClass->getName());` after [this line of code](https://github.com/Ocramius/ProxyManager/blob/master/src/ProxyManager/ProxyGenerator/NullObjectGenerator.php#L56), then my test case works, but then a few of the datasets cause [this test](https://github.com/Ocramius/ProxyManager/blob/master/tests/ProxyManagerTest/ProxyGenerator/NullObjectGeneratorTest.php#L57) to fail.

Am I on the right track at all? Or does the above give you enough information to resolve this fairly quickly? I'm not terribly familiar with ProxyManager's internals, so I'd rather check with you before digging any deeper.

**Comment by Ocramius** [comment]
I suggest writing a functional test case first: see [`tests/language-feature-scripts`](https://github.com/Ocramius/ProxyManager/tree/6e15877fc03cc211ce798d497a21208a44f7b726/tests/language-feature-scripts)

Also, please remember to use absolute references for files, as the branches keep moving, and the diffs change (and your links become invalid)

**Comment by Ocramius** [comment]
Corrected via #301. Rewording title though, as the return types were actually respected, but the inheritance wasn't.
